doc/developer/design/20230814_unbilled_replicas.md

```diff
@@ -0,0 +1,203 @@
+CREATE REPLICA SET cluster.support WITH (MANAGED = FALSE, BILLED AS 'unbilled');
+-- Create a replica in the replica set.
+CREATE CLUSTER REPLICA cluster.r1 SIZE '3xlarge' IN REPLICA SET support;
+-- Drop the replica set, also dropping replica `r1`.
+DROP REPLICA SET cluster.support CASCADE;
+```
+
+### Replica sets and replicas
+
+At the moment, clusters only contain replicas.
+With the introduction of replica sets, we change this property to allow different kinds of objects as part of a cluster.
+Replicas and replica sets share a common cluster namespace, so their names need to be unique within a cluster, but can be reused across clusters.
+
+At this time, users do not need to work with replica sets if they don't choose to.
+Each cluster has an implicit replica set (a.k.a, the `NULL` replica set?) that they interact with using the managed or unmanaged cluster SQL API.
+
+The following example demonstrates using the implicit replica set:
+```sql
+CREATE CLUSTER clname SIZE '3xlarge';
+```
+
+### Resources within replica sets
+
+Replicas are part of a replica sets, and their lifetime is tied to that of the replica set:
+
+```sql
+-- Cluster `cl` has an implicit replica set with one replica `r1`
+DROP REPLICA SET cl.default;
+ERROR: Replica set cl.default has objects
+HINT: Drop objects first, or specify CASCADE.
+
+DROP REPLICA SET cl.default CASCADE;
+```
+
+### Managed and unmanaged replica sets
+
+We can mix managed and unmanaged replica sets:
+
+```sql
+-- Create the a cluster with an implicit managed replica set, which include replicas `cl.r1`, `cl.r2`:
+CREATE CLUSTER cl SIZE '3xsmall', REPLICATION FACTOR 2;
+-- Create the `p1` replica set, which includes replica `cl.p1_r1`:
+CREATE REPLICA SET cl.p1 REPLICAS (p1_r1 SIZE 'large');
+```
+
+This requires a change to the names of replicas in managed clusters. Instead of
+only using the names `r1..rN`, we could use the replica set name, or `default`, as a
+prefix.
+
+The syntax to drop replicas stays the same:
+
+```sql
+-- Drop replica a replica:
+DROP CLUSTER REPLICA cl.r1;
+```
+
+### Altering replica sets
+
+* Similar to altering clusters
+  ```sql
+  ALTER REPLICA SET cl.p1 SET (REPLICATION FACTOR 0);
+  ```
+* Altering replica set name: Rename the replica set, rename replicas for managed replica sets
+  ```sql
+  ALTER REPLICA SET cl.p1 RENAME TO p2;
+  ```
+* Should we allow moving replicas between replica sets?
+  ```sql
+  ALTER CLUSTER REPLICA cl.r1 IN REPLICA SET support;
+  ERROR: Cannot move billed replica to unbilled replica set.
+  ```
+* Can we use RBAC to limit what users can do with unbilled replica sets?
+  * Owner can modify
+  * Owner can change owner
+  * Cluster owner can drop
+
+### System catalog
+
+We need to present replica sets in the system catalog.
+We extend the `mz_cluster_replicas` table with a `replica_set_id` column that names the replica set ID, or `NULL` if the replica is part of the implicit default replica set.
```

Comment:
I'm a bit worried about the `NULL` case, because I could imagine a future where users (or us) want to be able to do something with the implicit set and using `NULL` to designate that is real weird. Is there a const string we can choose instead? That clearly has problems too, because it gets funky if we ever allow re-naming of replica sets, then we have to figure out if the implicit set is still that const, it just doesn't exist, or if the implicitness follows through the rename. Hmm.

An analog here is the public schema which is auto created in postgres and many people don't know it exists or that schemas exist for lots of their postgres use. It's an invisible feature until you need it. I think we will help our future selves by not using NULL anywhere, but instead choosing some name that this is. That makes it easy to add an optional `IN REPLICA SET x` to all the cluster/replica commands that unwraps to this string if not present. Then no part of our internal logic needs to handle the None/NULL case. `"default"` is weird but not terrible. `"mz_default"`? Dunno.

Reply:
Agreed on your thought process. Also unsure what the right answer is. `default` is going to be hard because it will conflict with any replicas that people have created to date that are called `default`.